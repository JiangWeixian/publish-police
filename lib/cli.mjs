import cac from 'cac';
import { readJson } from 'fs-extra';
import { globby } from 'globby';

const DefaultIgnore = [
  ".git",
  "CVS",
  ".svn",
  ".hg",
  ".lock-wscript",
  ".wafpickle-N",
  ".*.swp",
  ".DS_Store",
  "._*",
  "npm-debug.log",
  ".npmrc",
  "node_modules",
  "config.gypi",
  "*.orig",
  "package-lock.json",
  "node_modules"
];
const glob = async (files, options = { cwd: process.cwd() }) => {
  const results = await Promise.all(files.map((file) => globby(file, {
    ignore: DefaultIgnore,
    ignoreFiles: [".npmignore"],
    gitignore: true,
    dot: true,
    ...options
  })));
  return results;
};
const distCheck = async ({
  strict = true,
  cwd = process.cwd()
}) => {
  let files = [];
  try {
    const config = await readJson(`${cwd}/package.json`);
    files = config.files ?? [];
  } catch (_) {
    throw new Error("package.json not found!");
  }
  if (files.length === 0 && strict) {
    throw new Error("files in package.json not found!");
  }
  if (files.length === 0 && !strict) {
    return true;
  }
  const results = await glob(files, { cwd });
  for (const [index, pattern] of files.entries()) {
    if (!results[index].length) {
      throw new Error(`\`${pattern}\` looks like empty or not exit! Maybe you add it in ignore files or build failed?`);
    }
  }
  return true;
};

const cli = cac();
cli.command("").option("--strict", "Choose a project type", {
  default: true
}).action((options) => {
  distCheck({ strict: options.strict });
});
cli.parse();
//# sourceMappingURL=cli.mjs.map
