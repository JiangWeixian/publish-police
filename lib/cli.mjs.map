{"version":3,"file":"cli.mjs","sources":["../src/lib.ts","../src/cli.ts"],"sourcesContent":["import { readJson } from 'fs-extra'\nimport type { PackageJson } from 'type-fest'\nimport { globby } from 'globby'\n\ntype Options = {\n  cwd?: string\n}\n\n// https://docs.npmjs.com/cli/v8/configuring-npm/package-json#files\nconst DefaultIgnore = [\n  '.git',\n  'CVS',\n  '.svn',\n  '.hg',\n  '.lock-wscript',\n  '.wafpickle-N',\n  '.*.swp',\n  '.DS_Store',\n  '._*',\n  'npm-debug.log',\n  '.npmrc',\n  'node_modules',\n  'config.gypi',\n  '*.orig',\n  'package-lock.json',\n  'node_modules',\n]\n\nexport const glob = async (files: string[], options: Options = { cwd: process.cwd() }) => {\n  const results = await Promise.all(\n    files.map((file) =>\n      globby(file, {\n        ignore: DefaultIgnore,\n        ignoreFiles: ['.npmignore'],\n        gitignore: true,\n        dot: true,\n        ...options,\n      }),\n    ),\n  )\n  return results\n}\n\nexport const distCheck = async ({\n  strict = true,\n  cwd = process.cwd(),\n}: Options & { strict?: boolean }) => {\n  let files: string[] = []\n  try {\n    const config: PackageJson = await readJson(`${cwd}/package.json`)\n    files = config.files ?? []\n  } catch (_) {\n    throw new Error('package.json not found!')\n  }\n  if (files.length === 0 && strict) {\n    throw new Error('files in package.json not found!')\n  }\n  // in non-strict mode, empty files is allowed\n  // npm will always upload files in current directory\n  if (files.length === 0 && !strict) {\n    return true\n  }\n  const results = await glob(files, { cwd })\n  for (const [index, pattern] of files.entries()) {\n    if (!results[index].length) {\n      throw new Error(\n        `\\`${pattern}\\` looks like empty or not exit! Maybe you add it in ignore files or build failed?`,\n      )\n    }\n  }\n  return true\n}\n","import cac from 'cac'\nimport { distCheck } from './lib'\n\nconst cli = cac()\n\ncli\n  .command('')\n  .option('--strict', 'Choose a project type', {\n    default: true,\n  })\n  .action((options) => {\n    distCheck({ strict: options.strict })\n  })\n\ncli.parse()\n"],"names":[],"mappings":";;;;AAEA,MAAM,aAAa,GAAG;AACtB,EAAE,MAAM;AACR,EAAE,KAAK;AACP,EAAE,MAAM;AACR,EAAE,KAAK;AACP,EAAE,eAAe;AACjB,EAAE,cAAc;AAChB,EAAE,QAAQ;AACV,EAAE,WAAW;AACb,EAAE,KAAK;AACP,EAAE,eAAe;AACjB,EAAE,QAAQ;AACV,EAAE,cAAc;AAChB,EAAE,aAAa;AACf,EAAE,QAAQ;AACV,EAAE,mBAAmB;AACrB,EAAE,cAAc;AAChB,CAAC,CAAC;AACK,MAAM,IAAI,GAAG,OAAO,KAAK,EAAE,OAAO,GAAG,EAAE,GAAG,EAAE,OAAO,CAAC,GAAG,EAAE,EAAE,KAAK;AACvE,EAAE,MAAM,OAAO,GAAG,MAAM,OAAO,CAAC,GAAG,CAAC,KAAK,CAAC,GAAG,CAAC,CAAC,IAAI,KAAK,MAAM,CAAC,IAAI,EAAE;AACrE,IAAI,MAAM,EAAE,aAAa;AACzB,IAAI,WAAW,EAAE,CAAC,YAAY,CAAC;AAC/B,IAAI,SAAS,EAAE,IAAI;AACnB,IAAI,GAAG,EAAE,IAAI;AACb,IAAI,GAAG,OAAO;AACd,GAAG,CAAC,CAAC,CAAC,CAAC;AACP,EAAE,OAAO,OAAO,CAAC;AACjB,CAAC,CAAC;AACK,MAAM,SAAS,GAAG,OAAO;AAChC,EAAE,MAAM,GAAG,IAAI;AACf,EAAE,GAAG,GAAG,OAAO,CAAC,GAAG,EAAE;AACrB,CAAC,KAAK;AACN,EAAE,IAAI,KAAK,GAAG,EAAE,CAAC;AACjB,EAAE,IAAI;AACN,IAAI,MAAM,MAAM,GAAG,MAAM,QAAQ,CAAC,CAAC,EAAE,GAAG,CAAC,aAAa,CAAC,CAAC,CAAC;AACzD,IAAI,KAAK,GAAG,MAAM,CAAC,KAAK,IAAI,EAAE,CAAC;AAC/B,GAAG,CAAC,OAAO,CAAC,EAAE;AACd,IAAI,MAAM,IAAI,KAAK,CAAC,yBAAyB,CAAC,CAAC;AAC/C,GAAG;AACH,EAAE,IAAI,KAAK,CAAC,MAAM,KAAK,CAAC,IAAI,MAAM,EAAE;AACpC,IAAI,MAAM,IAAI,KAAK,CAAC,kCAAkC,CAAC,CAAC;AACxD,GAAG;AACH,EAAE,IAAI,KAAK,CAAC,MAAM,KAAK,CAAC,IAAI,CAAC,MAAM,EAAE;AACrC,IAAI,OAAO,IAAI,CAAC;AAChB,GAAG;AACH,EAAE,MAAM,OAAO,GAAG,MAAM,IAAI,CAAC,KAAK,EAAE,EAAE,GAAG,EAAE,CAAC,CAAC;AAC7C,EAAE,KAAK,MAAM,CAAC,KAAK,EAAE,OAAO,CAAC,IAAI,KAAK,CAAC,OAAO,EAAE,EAAE;AAClD,IAAI,IAAI,CAAC,OAAO,CAAC,KAAK,CAAC,CAAC,MAAM,EAAE;AAChC,MAAM,MAAM,IAAI,KAAK,CAAC,CAAC,EAAE,EAAE,OAAO,CAAC,kFAAkF,CAAC,CAAC,CAAC;AACxH,KAAK;AACL,GAAG;AACH,EAAE,OAAO,IAAI,CAAC;AACd,CAAC;;ACpDD,MAAM,GAAG,GAAG,GAAG,EAAE,CAAC;AAClB,GAAG,CAAC,OAAO,CAAC,EAAE,CAAC,CAAC,MAAM,CAAC,UAAU,EAAE,uBAAuB,EAAE;AAC5D,EAAE,OAAO,EAAE,IAAI;AACf,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC,OAAO,KAAK;AACvB,EAAE,SAAS,CAAC,EAAE,MAAM,EAAE,OAAO,CAAC,MAAM,EAAE,CAAC,CAAC;AACxC,CAAC,CAAC,CAAC;AACH,GAAG,CAAC,KAAK,EAAE"}