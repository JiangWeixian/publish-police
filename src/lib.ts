import { readFileSync } from 'fs'
import type { PackageJson } from 'type-fest'
import { globby } from 'globby'

type Options = {
  cwd?: string
}

// https://docs.npmjs.com/cli/v8/configuring-npm/package-json#files
const DefaultIgnore = [
  '.git',
  'CVS',
  '.svn',
  '.hg',
  '.lock-wscript',
  '.wafpickle-N',
  '.*.swp',
  '.DS_Store',
  '._*',
  'npm-debug.log',
  '.npmrc',
  'node_modules',
  'config.gypi',
  '*.orig',
  'package-lock.json',
  'node_modules',
]

export const glob = async (files: string[], options: Options = { cwd: process.cwd() }) => {
  const results = await Promise.all(
    files.map((file) =>
      globby(file, {
        ignore: DefaultIgnore,
        ignoreFiles: ['.npmignore'],
        gitignore: true,
        dot: true,
        ...options,
      }),
    ),
  )
  return results
}

export const distCheck = async ({
  strict = true,
  cwd = process.cwd(),
}: Options & { strict?: boolean }) => {
  let files: string[] = []
  try {
    const config: PackageJson = JSON.parse(readFileSync(`${cwd}/package.json`).toString('utf-8'))
    files = config.files ?? []
  } catch (_) {
    throw new Error('package.json not found!')
  }
  if (files.length === 0 && strict) {
    throw new Error('files in package.json not found!')
  }
  // in non-strict mode, empty files is allowed
  // npm will always upload files in current directory
  if (files.length === 0 && !strict) {
    return true
  }
  const results = await glob(files, { cwd })
  for (const [index, pattern] of files.entries()) {
    if (!results[index].length) {
      throw new Error(
        `\`${pattern}\` looks like empty or not exit! Maybe you add it in ignore files or build failed?`,
      )
    }
  }
  return true
}
