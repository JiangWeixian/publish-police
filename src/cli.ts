import cac from 'cac'
import { filesCheck, mainCheck, exportsCheck, readPkgJson } from './lib'
import Listr from 'listr'
import colors from 'picocolors'
// @ts-expect-error -- ignore types
import mlistr from 'listr-multiline-renderer'

const cli = cac('publish-police')

cli
  .command('')
  .option('--strict', '[boolean] Strict mode', {
    default: true,
  })
  .action(async (options) => {
    try {
      const pkg = readPkgJson(process.cwd())
      const tasks = new Listr(
        [
          {
            title: `Check ${colors.cyan('`files`')}`,
            task: () => filesCheck({ strict: options.strict, pkg }),
          },
          {
            title: `Check ${colors.cyan('`main and module`')}`,
            task: () => mainCheck({ pkg }),
          },
          { title: `Check ${colors.cyan('`exports`')}`, task: () => exportsCheck({ pkg }) },
        ],
        { renderer: mlistr },
      )
      tasks.run().catch(() => {
        process.exit(1)
      })
    } catch (e) {
      process.exit(0)
    }
  })

cli.parse()
