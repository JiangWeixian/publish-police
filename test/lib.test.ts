import { glob, distCheck } from '../src/lib'
import { describe, it, expect } from 'vitest'
import path from 'path'

describe('glob', () => {
  it('non exit files should return empty array', async () => {
    const results = await glob(['dist'])
    expect(results.length).toBe(1)
    expect(results[0].length).toBe(0)
  })

  it('exit files should return files list', async () => {
    const results = await glob(['bin'])
    expect(results.length).toBe(1)
    expect(results[0]).toMatchInlineSnapshot(`
    [
      "bin/index.mjs",
    ]
    `)
  })

  it('cwd should work', async () => {
    const results = await glob(['lib'], { cwd: path.resolve(__dirname, './fixtures') })
    expect(results[0]).toMatchInlineSnapshot(`
    [
      "lib/.gitkeep",
    ]
    `)
  })

  it('default ignores should work', async () => {
    const results = await glob(['node_modules'])
    expect(results.length).toBe(1)
    expect(results[0].length).toBe(0)
  })

  it('.npmignore should work', async () => {
    const results = await glob(['.changeset'])
    expect(results.length).toBe(1)
    expect(results[0].length).toBe(0)
  })
})

describe('dist-checker', () => {
  it('non-strict mode should work', async () => {
    expect(await distCheck({ strict: false })).toBe(true)
  })

  it('strict mode should work', async () => {
    expect(async () => await distCheck({ strict: true })).to.rejects.toThrow(
      'files in package.json not found!',
    )
  })

  it('should work', async () => {
    expect(await distCheck({ strict: false, cwd: path.resolve(__dirname, './fixtures') })).toBe(
      true,
    )
  })
})
