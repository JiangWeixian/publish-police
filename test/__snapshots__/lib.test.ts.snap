// Vitest Snapshot v1

exports[`resolve options > resolve options 1`] = `
{
  "files": [
    [
      "lib/.gitkeep",
    ],
    [
      "exports/test.js",
    ],
  ],
  "packageJson": {
    "exports": {
      ".": "./lib/cli.mjs",
      "./a": {
        "node": {
          "require": "./exports/test.js",
        },
      },
      "./b": "./exports/test.js",
    },
    "files": [
      "lib",
      "exports",
    ],
  },
  "root": "<rootDir>",
  "strict": true,
}
`;
